wasm/jit/engine.go

```diff
@@ -0,0 +1,329 @@
+	jitStatusCallHostFunction
+	// TODO: trap, etc?
+)
+
+var (
+	engineStackSliceOffset              = int64(unsafe.Offsetof((&engine{}).stack))
+	engineCurrentStackPointerOffset     = int64(unsafe.Offsetof((&engine{}).currentStackPointer))
+	engineCurrentBaseStackPointerOffset = int64(unsafe.Offsetof((&engine{}).currentBaseStackPointer))
+	engineJITStatusOffset               = int64(unsafe.Offsetof((&engine{}).jitCallStatusCode))
+	engineFunctionCallIndexOffset       = int64(unsafe.Offsetof((&engine{}).functionCallIndex))
+	engineContinuationAddressOffset     = int64(unsafe.Offsetof((&engine{}).continuationAddressOffset))
+)
+
+type callFrame struct {
+	continuationAddress      uintptr
+	continuationStackPointer uint64
+	baseStackPointer         uint64
+	f                        *compiledWasmFunction
+	caller                   *callFrame
+}
+
+func (c *callFrame) String() string {
+	return fmt.Sprintf(
+		"[continuation address=%d, continuation stack poitner=%d, base stack pointer=%d]",
+		c.continuationAddress, c.continuationStackPointer, c.baseStackPointer,
+	)
+}
+
+type compiledWasmFunction struct {
+	inputNum, outputNum uint64
+	codeSegment         []byte
+	memoryInst          *wasm.MemoryInstance
+	codeInitialAddress  uintptr
+	memoryAddress       uintptr
+}
+
+const (
+	builtinFunctionIndexGrowMemory = iota
+)
+
+func (e *engine) stackGrow() {
+	newStack := make([]uint64, len(e.stack)*2)
+	copy(newStack[:len(e.stack)], e.stack)
+	e.stack = newStack
+}
+
+func (e *engine) exec(f *compiledWasmFunction) {
+	e.callFrameStack = &callFrame{
+		continuationAddress:      f.codeInitialAddress,
+		f:                        f,
+		caller:                   nil,
+		continuationStackPointer: f.inputNum,
+	}
+	// TODO: We should check the size of the stack,
+	// and if it's running out, grow it before calling into JITed code.
+	// It should be possible to check the necessity by statically
+	// analyzing the max height of the stack in the function.
+	if false {
+		e.stackGrow()
+	}
+	for e.callFrameStack != nil {
+		currentFrame := e.callFrameStack
+		if buildoptions.IsDebugMode {
+			fmt.Printf("callframe=%s, currentBaseStackPointer: %d, currentStackPointer: %d, stack: %v\n",
+				currentFrame.String(), e.currentBaseStackPointer, e.currentStackPointer,
+				e.stack[:e.currentBaseStackPointer+e.currentStackPointer],
+			)
+		}
+
+		// Call into the jitted code.
+		jitcall(
+			currentFrame.continuationAddress,
+			uintptr(unsafe.Pointer(e)),
+			currentFrame.f.memoryAddress,
+		)
+
+		// Check the status code from JIT code.
+		switch e.jitCallStatusCode {
+		case jitStatusReturned:
+			// Meaning that the current frame exits
+			// so we just get back to the caller's frame.
+			callerFrame := currentFrame.caller
+			e.callFrameStack = callerFrame
+			if callerFrame != nil {
+				e.currentBaseStackPointer = callerFrame.baseStackPointer
+				e.currentStackPointer = callerFrame.continuationStackPointer
+			}
+		case jitStatusCallWasmFunction:
```

Comment:
this feels more like states than status? jitStateCallWasmFunction?

Reply:
it's not state, but the status of jitcall. Will explain in the doc comment on jitCallStatusCode type